### DISCUSSION TITLE
Allow docker to not bind all interfaces when following the reverse proxy documentation

### DISCUSSION THREAD

**Discussion by wabberz** [nextcloud/all-in-one#883] [Ideas] [answered]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1. Follow [reverse-proxy.md](https://github.com/nextcloud/all-in-one/blob/main/reverse-proxy.md)
2. `netstat -tulpen` or `ss -lunt`

### Expected behavior 
Being a _reverse proxy setup_ I don't expect it to bind to all interfaces but only the local loopback so the reverse proxy can decide which ports to open to the world. This is a major security concern for people who just set up ufw or firewalld without an external firewall since docker will just sidestep them and happily allow connections from outside.

### Actual behavior 
This setup binds to:

- 0.0.0.0:11000/tcp
- 0.0.0.0:3478/tcp+udp
- :::11000/tcp
- :::3478/tcp+udp

### Host OS 
Arch x64 on bare metal

#### Nextcloud AIO version 
1.4.2

#### Current channel 
latest

**Comment by szaimen** [comment]
This is a feature request hence moving to discussions

**Comment by szaimen** [comment]
See https://github.com/nextcloud/all-in-one/discussions/731#discussioncomment-2841409 for a workaround

**Reply by wabberz** [reply]
> This is a feature request hence moving to discussions

I'm not sure that's true, I do consider it a bug. It's bad practice to bind to all interfaces if the user doesn't explicitly ask for it, especially since Docker sidesteps local firewalls and not everyone seems to be aware of that. There's tons of docker setups with ports open that shouldn't be because everyone expects their local firewall to protect them. My hoster only provides an external IPv4 firewall which means I'm SOL on IPv6 ports that bind to :::.

It shouldn't just be a feature the user can turn on but the default imho.

**Reply by szaimen** [reply]
So the workaround does not work for you?

**Reply by wabberz** [reply]
I was referring to the removal of the bug label. The workaround does indeed work, but it is just that - a workaround. Not everyone runs portscans on their own devices and a shocking number of Docker users is not aware that Docker sidesteps firewalls.

Of course that's their responsibility, but default settings should be safe imho.

**Reply by szaimen** [reply]
I'll add the workaround to the reverse proxy documentation 👍

**Reply by wabberz** [reply]
That's good as a warning, but I think it would be much better to only bind to the local loopback. There is no good reason to bind to all interfaces in this scenario, is there?

**Reply by szaimen** [reply]
There is for cases where the reverse proxy is not running on the same server

**Reply by wabberz** [reply]
I think the user should explicitly say they want to bind on all interfaces in this scenario. Even if the reverse proxy is not running on the same server, they may not want to bind to all interfaces but only a specific one. They could very well have interfaces that shouldn't be touched at all by this.

**Reply by szaimen** [reply]
I am interested: How would you limit publishing the port to a specific interface if this was a docker run command?
Or are you referring to docker networks?

**Reply by wabberz** [reply]
`-p 11000:11000`
would be
`-p 127.0.0.1:11000:11000`

**Reply by szaimen** [reply]
I see. Thanks for clarifying! In this case the workaround seems like a valid approach to me as you reach the same goal by running the iptables command, afaics.

**Reply by wabberz** [reply]
Well, you may have other interfaces. Say your server is in an internal network and you only want to make the service available to a specific subnet, you could do:

`-p 10.1.2.3:11000:11000`

Of course you can do a workaround, but you shouldn't bind to all interfaces if you don't need to in the first place. It's just bad practice imho.

**Reply by szaimen** [reply]
It may be bad practise but definitely much easier to configure for the user as he does not need to care about subnets, etc.

**Reply by wabberz** [reply]
But they do end up with open ports they didn't intend to have open if they're unaware of Docker's behavior. A lot of people blindly trust their local firewall. I consider this a major security concern, since it's not even possible to change the interface the services bind to without entering the container and rewriting some scripts.

**Reply by szaimen** [reply]
Yeah but if we document this behaviour and the workaround you can change it if you need to.

**Reply by wabberz** [reply]
So it doesn't need to be safe by default as long as it works? I think I should look into other solutions if this practice is common around here... Thanks for your time!

**Reply by szaimen** [reply]
No, it is safe if you configure iptables correctly. I do not see any problem with that.

You are welcome! Thanks for your time, too!

**Reply by wabberz** [reply]
**If**. If some other service screws up your iptables, if you haven't made the change permanent (common with people new to Linux), if something doesn't work and the user disables the firewall altogeter (dumb, but I've seen it happen), ...

It's just the principle of least amount of privileges necessary to run the service so it doesn't fall apart when one layer fails. Of course you're right in that the user needs to practice good security on their own, but reality is they don't.

**Reply by szaimen** [reply]
That is why I will document how to configure it correctly in the same documentation where is documented how to to run it behind a reverse proxy.

**Comment by szaimen** [comment]
I will not work on adding this functionality as the workaround (that I will document with https://github.com/nextcloud/all-in-one/issues/884) should be sufficient in most cases. However PRs are welcome!

**Comment by szaimen** [comment, marked as answer]
I will add this with https://github.com/nextcloud/all-in-one/issues/889

**Reply by szaimen** [reply]
This is now released with 1.6.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel
